### ISSUE TITLE
`QueryExecutionContext.Filter` causing unnecessary overhead for read-only contexts with large indexes

### ISSUE THREAD

**Issue by mattbab** [chriseldredge/Lucene.Net.Linq#36]
It appears that using a catch-all filter on `LuceneQueryExecutorBase.ExecuteCollection(136)` causes 100-400 ms of unnecessary overhead in `IndexSearcher.Search` when the context is read-only and working with > 50K records/index. This is caused by the `QueryExecutionContext.Filter` being set to `<KeyField>:*` and filtering for ALL records when not necessary.

I can't tell whether this was a necessary adjustment to get your Nuget implementation going or not, but based on the documentation, this filter is only necessary when Updates/Deletes are possible. This should never be the case when in a read-only context so the performance degradation shouldn't apply.

As a dirty workaround, I've commented out `QueryModelTranslator.cs(99)` so the `Filter` property is never set, but this might break whatever use case this was initially intended to address (which likely doesn't apply to my environment). I couldn't find a good way in your code to only turn this off for searches (though I didn't try that hard either).

If this is necessary for your purposes then I can fork the project or help you work through the test scenario. Otherwise, let me know if you need a repro or pull request for further detail.

Thanks.

**Comment by chriseldredge** [comment]
Imagine two different entities are stored in the index:

``` c#
public class Article
{
    [Field(Key = true)]
    public long ArticleId { get; set; }

    public string Title { get; set; }
}

public class Comment
{
    [Field(Key = true)]
    public long CommentId { get; set; }

    public string Title { get; set; }
}
```

And you want to retrieve articles by title. Without adding the filter query `ArticleId:*`, some comments that match the query on Title could be returned, leading later to cast exceptions and other undesired behavior. This is the intention of the design: to support indexing and retrieval of different entity types that may have shared field names.

In the case where only one type of entity is being stored, or extreme care has been taken to ensure each entity indexes unique field names, I see how adding this filter query could slow down performance with no benefit.

Perhaps one solution would be to have an advanced setting that allows clients to disable this behavior.

**Comment by mattbab** [comment]
Even in the provided scenario, if either entity has a significant number of entities you end up paying an enormous inherent cost with little advantage over following Lucene best practices and keeping your entities in separate indexes. I've been designing an extensible and easy-to-use infrastructure to support this design around your library for this specific purpose.

Your proposed solution sounds functional, though it seems strange that you'd want the library to prefer the design pattern that requires additional overhead for the simplest use case. Even worse, the overhead of supporting such pattern is an order of magnitude worse for even somewhat modest sized indexes (100,000+).

I just wonder if it wouldn't be more advantageous to define best practices for scenarios that require additional overhead than imply the overhead on even the simplest test index.

**Comment by chriseldredge** [comment]
I was not aware that there was a heavy performance cost associated with such filter queries. After reusing the same index of IndexSearcher (#35) do you still experience such a penalty or does the field cache improve performance after the searcher is warm? If you still get terrible performance when using a warm IndexSearcher I would reconsider the use of filter queries for this purpose.

**Comment by mattbab** [comment]
My tests were primarily with an identical query on a single entity definition (albeit several thousand times) and performance never seems to go below 300 ms/query (15 total fields with 10 analyzed on ~2 million documents). With my hack, I easily get under 30 ms even on warm up.

**Comment by chriseldredge** [comment]
I was able to reproduce this, and committed 2829e6d272814e614f7e6a3e137308e32df66805 to allow clients to disable it. I agree this should be the default behavior, but don't want to break existing clients. You can enable the optimization by doing:

```
luceneDataProvider.Settings.EnableMultipleEntities = false;
```

I'll close this issue after releasing to nuget.org.

**Comment by chriseldredge** [comment]
Released in 3.2.53.
